Various fixes of varying importance

DalekCraft2 opened the pull request:
I would list the fixes, but they are sort of listed already in the commits. I will clarify, though, that I migrated from `WebClient` to `HttpClient` because the former is obsolete.

I also have an incoming second PR which upgrades the project to .NET 7.0 (and changes the project's version format to be like "1.0.3" instead of "1.03"), which I separated from this one because that would be too big of a change to lump in with these relatively smaller changes (and because it includes these changes anyway, so you would need to merge this one first).

(while i'm at it, i recommend not using periods to separate words in the file names when releasing on github, but they can still be used for separating the version components)

David-JonesDVN committed:
Merge pull request #11 from David-JonesDVN/dev

Pre-release 4 is now final, with the completion of full support for thcrap.

David-JonesDVN committed:
Merge branch 'master' of https://github.com/David-JonesDVN/Touhou-Relauncher

David-JonesDVN committed:
Merge pull request #15 from David-JonesDVN/dev

1.02 Release

David-JonesDVN committed:
Merge pull request #18 from David-JonesDVN/dev

Touhou 19 support (1.03)

DalekCraft2 committed:
Make default starting repo the same as the one used on first startup

DalekCraft2 committed:
Fix exception from invalid thcrap path

DalekCraft2 committed:
Update to .NET Framework 4.8.1

Might end up switching to .NET Core later, though.

DalekCraft2 committed:
Run formatter everywhere (except auto-generated files)

DalekCraft2 committed:
Optimize imports

DalekCraft2 committed:
Check for keys existing before adding them in `thcrap.cs`

DalekCraft2 committed:
Migrate from WebClient to HttpClient, and pass owner arg to ShowDialog calls

DalekCraft2 committed:
Remove unused imports

DalekCraft2 committed:
Fix errors and remove unused references

DalekCraft2 committed:
Use a FolderBrowserDialog for crapDir because it saves the value as a directory

DalekCraft2 committed:
Have all file/folder browsers start at their previous config values

DalekCraft2 committed:
Replace appspot button with a website which still exists

DalekCraft2 committed:
Fix formatting of game numbers

DalekCraft2 committed:
Change a constant from 9 to 10

Judging from the lines below this one referring to th10, I'd say this was supposed to be a 10, and not a 9.

DalekCraft2 committed:
Update ConfigForm.cs

DalekCraft2 committed:
Update README.md

DalekCraft2 committed:
Update README.md

DalekCraft2 committed:
Use the GitHub VS .gitignore template

David-JonesDVN commented:
Hello, thanks for the fixes. I've barely managed to look through the code since last year, so I appreciate the help. I'll be looking through the changes when I find the time, and commenting on them if I have any questions.
Regarding the .NET Framework version, I've kept it at 4.8 to retain backwards compatibility with Windows 7 according to Microsoft's compatibility list, and I haven't felt the need to upgrade to something newer.

David-JonesDVN reviewed (commented):
This (91b6577) feels like too much complexity for constants with no use besides implementation details, considering the formatted numbers are the only ones used in code. Was this change made as a prerequisite for planned functionality?
(I have no idea why the review ended up here. I posted it on the commit itself. Shows how much I use Github.)

DalekCraft2 commented:
Previously, the `String.Format("00")` call did not work properly when formatting the game numbers of some spinoffs (e.g., th075's number would be formatted to "75" instead of "075"), which caused a lot of code related to those strings to not work properly (e.g., code for searching for th075's executables would be searching for "th75.exe" instead of "th075.exe").

David-JonesDVN commented:
I see. In that case, would it be better to switch to strings then? IIRC the reason I made the number array to begin with was to get the game indices and to use them as suffixes, neither of which actually need the elements to be numbers.

DalekCraft2 commented:
I would say yes, if not for the replay downloader. It doesn't use strings like "095"; it uses strings like "95" because the replay files have names like that, unlike everything else. It might be easier to keep the array as a numbers array for that.

By the way, I just noticed this; why is `nameToID` a dictionary instead of a list? With how it's set up, one could just convert it to a list and check the index of one of the entries instead of assigning an index to each key.

~~Actually, better idea: use a `SortedDictionary` with the game numbers as the keys and the game names as the values. That way, `nameToID` and `idToNumber` can be combined into one field.~~

actually, never mind that last idea; it seems that `SortedDictionary` does not have a method for indexing, and forces its keys to be in a certain order; i am sort of new to C#

David-JonesDVN commented:
Ah right, replays. Still, I'll try to see if I can come up with something cleaner for game suffixes (for example, switching to strings while in the replay downloader checking the suffix both with and without a leading zero, with the assumption that there won't be 75 mainline games). If not, we'll go with this.
About `nameToID`, I think that was due to me not wanting to use `IndexOf` where I could avoid it. I'd definitely say that was unwarranted. I was planning to merge the two into a list of `<string, int>` pairs, but back then I didn't know how to get indices from that. I think it can be done using List's `FindIndex` method, but even then, I don't think merging the lists/using index lookups make the code look any nicer. I'm leaving this part to your discretion.

DalekCraft2 committed:
Change back to .NET Framework 4.8 for Win7 compatibility

DalekCraft2 committed:
Comment out an outdated NP2-related line of code

Keeping it as a comment in case np2Dir gets stored as a directory again in the future.

DalekCraft2 committed:
Refactor some dictionaries to lists

DalekCraft2 committed:
Refactor the names of three fields for clarification

DalekCraft2 commented:
May I also ask why you always release two versions, one with the thcrap configurator and one without it? I do not see what the point of doing that is.

DalekCraft2 committed:
Remove unused Icon_th0951 resource

DalekCraft2 commented:
i'd also like to bring to attention that the `trayIcon` element in `MainForm.Designer` seems to not be visible, making its context menu inaccessible

David-JonesDVN commented:
I know some people who really dislike thcrap and any integration with it and would prefer using static patches/custom thcrap setups for each game wherever possible. The offline version is made with those people in mind.
The tray icon is disabled by default. Its visibility is set in `MainForm_Show`, according to the configuration file.
Also, regarding the `NekoProject` function, it's there to change the HDI path in Neko Project II's configuration file, and return an error if it can't. I guess I forgot a file check before reading the ini file, and there might be better ways to use exception handling in it, but IIRC I always stored np2Dir as a file path.

David-JonesDVN commented on Touhou Launcher/MainForm.cs:
```diff
@@ -1039,25 +1039,31 @@ private void chkRandom_CheckedChanged(object sender, EventArgs e)
-            TextBox txtbox = (TextBox)launcherSettings.Controls.Find(((Button)sender).Name.Substring(6).ToLower() + "Dir", false).FirstOrDefault(n => n.GetType() == typeof(TextBox));
-            foreach (string file in MainForm.FileBrowser(this, MainForm.rm.GetString(((Button)sender).Name.Substring(6).ToLower() + "SelectTitle"), MainForm.rm.GetString("executableFilter") + " (*.exe, *.bat, *.lnk)|*.exe;*.bat;*.lnk|" + MainForm.rm.GetString("allFilter") + " (*.*)|*.*"))
+            string controlName = ((Button)sender).Name.Substring(6).ToLower() + "Dir";
+            FieldInfo field = curCfg.GetType().GetField(controlName);
```
Could just be my brainworms speaking, but wouldn't it be better to use `typeof(Configs)` here? `curCfg`'s type isn't going to change at this point, so having the type at compile time is preferable. Also, having the program throw an easily noticeable exception in case of a type mismatch is preferable to failing silently.

DalekCraft2 replied:
Yeah, that makes sense.

DalekCraft2 committed:
Use typeof(Configs) instead of curCfg.GetType()

DalekCraft2 commented:
Regarding `NekoProject()`, I just realized that it uses `Path.GetDirectoryName(curCfg.np2Dir)`, so it *does* use the directory. However, it checks for `np21nt.ini`, which might not always be the right file if the user has set their executable to be one other than `np21nt.exe`. I will make it check for the right `.ini` file.

DalekCraft2 committed:
Uncomment NP2 code and fix bug with reading ini file

DalekCraft2 commented:
The program crashed when trying to read the `.ini` file if it did not exist, so I fixed that. It requires that the user runs the executable outside of the program now first, though, to ensure that the `.ini` file is created. Otherwise, they get the "configuration ini is not valid" message.

David-JonesDVN commented on Touhou Launcher/Touhou Launcher.csproj:
```diff
@@ -137,8 +126,6 @@
     </Compile>
   </ItemGroup>
   <ItemGroup>
-    <EmbeddedResource Include="jsonnet\Newtonsoft.Json.dll" />
```
Does `Newtonsoft.Json` have a statically linked library now? The lack of one was why I used this hack to include the dynamically linked library in the executable, as I'd much rather the executable be standalone.

DalekCraft2 replied:
If you are asking whether it has a NuGet package, yes.

David-JonesDVN replied:
Static library as in, a library that's linked at compile time and doesn't need to be included with the executable as a dll file.

DalekCraft2 replied:
Well, no, but .NET 6.0 allows single-file publishing, and it works on Windows 7, so I could migrate the project to that.

David-JonesDVN replied:
I see. In that case I'd recommend reverting the change for `Newtonsoft.Json` here and simply embed an updated dll for now. The port to another platform can wait until another pull request.

DalekCraft2 committed:
Have JSON.NET included the old way so it can be embedded

DalekCraft2 commented:
Late responses to [this](https://github.com/David-JonesDVN/Touhou-Relauncher/pull/19#issuecomment-1722220442):
* ~~For the tray icon, where does it show up? The only thing what happens when I double-click the window icon is the window's closing, so I don't think it's there (Clicking once reveals a separate, unrelated context menu which looks to contain simple Windows-related stuff).~~ Oh, wait, I just remembered what the tray is. It's the small icon set in the bottom right of the desktop.
* For the thcrap integration, can't the users who don't like thcrap just not use the thcrap features in the main version?

DalekCraft2 commented:
Would you like to use [Costura.Fody](https://github.com/Fody/Costura) to merge the Newtonsoft.Json DLL instead of using the current hack? I've tested it with the project, and it works.

David-JonesDVN commented:
I'd like to avoid extra dependencies if possible. Costura (and by extension Fody) does too much compared to this one small use case. In addition, it appears Costura uses the same method of hijacking the assembly resolver event anyway, so I don't think much is gained by using it.

DalekCraft2 committed:
Simplify unnecessarily fully qualified class names

DalekCraft2 committed:
Allow drag/drop with `MainForm` file path text fields

DalekCraft2 committed:
Simplify `string` and `char` names

DalekCraft2 committed:
Update Program.cs

DalekCraft2 committed:
Make `MainForm.GetAll()` generic

David-JonesDVN reviewed (approved):
Alright, I've gone through the changes and they all look good to me.
If there are still things you're working on, feel free to keep pushing them. Just let me know when you feel like there isn't much else you'd like to change, and I'll squash/merge after testing the build.

DalekCraft2 committed:
Fix FileBrowser initialDirectory crash if existing config value is empty string

DalekCraft2 committed:
Change incorrect usages of "directory" to fit context

I did not change anything related to the settings because that would break compatibility.

DalekCraft2 committed:
Remove broken link from info tab

DalekCraft2 committed:
Add `StartupObject` to `Touhou Launcher.csproj`

DalekCraft2 committed:
Update Touhou Launcher.csproj

DalekCraft2 committed:
Change the version format

Trust me, this is better.

DalekCraft2 committed:
Use "ThXX"-type IDs instead of game name acronyms

This way, we do not have to manually update `gameNames` for each new game, and the resources appear sorted by game order in the resource designers.

DalekCraft2 committed:
Update MainForm.Designer.cs

DalekCraft2 committed:
Use `using` blocks for some more dialogs

DalekCraft2 committed:
Use object initializers

DalekCraft2 committed:
Use null propagation in one spot

DalekCraft2 committed:
Use a constant for the number of PC-98 games

DalekCraft2 committed:
Use DPI-awareness

This fixes the blurry text. I'm going to need to rely on others to test it on other operating systems, though, because I only have Windows 11.

DalekCraft2 committed:
Update MainForm.Designer.cs

DalekCraft2 committed:
Update MainForm.Designer.cs

DalekCraft2 committed:
Disable DPI-awareness because it breaks the replay browser's anchors

DalekCraft2 committed:
Remove unnecessary line from `.csproj` file

DalekCraft2 committed:
Use `packages.config`

This was done for the sake of the next commit. It also makes this repo slightly more organized by using built-in features of this project format.

DalekCraft2 committed:
Migrate from `WebBrowser` to `WebView2` which is infinitely better

This fixes all of the script error popups and makes the pages appear as they would in a proper web browser.

DalekCraft2 commented:
You can test the build now. The remaining things what I want to add should probably go in a separate PR because they would involve:
* Migrating to an SDK-style project (This would make the `.csproj` file much, _much_ easier to read)
* Migrating from WinForms to WPF because the latter is newer, apparently fixes bugs from the former, and apparently has better DPI-awareness support (DPI-awareness would fix the blurry text, but it's kind of broken on WinForms because it breaks the replay browser's anchors)
* Upgrading to .NET 6 so single-file assemblies can be created without hacks (.NET 6 specifically because using a newer version would mean losing compatibility with Windows 7)

DalekCraft2 committed:
Remove a comment about how bad the .NET web browser is

We don't use that browser anymore, after all.

David-JonesDVN commented on Touhou Launcher/Touhou Launcher.csproj:
```diff
@@ -140,7 +141,7 @@
   </ItemGroup>
   <ItemGroup>
-    <EmbeddedResource Include="jsonnet\Newtonsoft.Json.dll" />
+    <EmbeddedResource Include="..\packages\Newtonsoft.Json.13.0.3\lib\net45\Newtonsoft.Json.dll" />
```
Since you're using the resource manager to embed the assemblies now, why are they included here again? I removed this and the executable ran perfectly fine without the library or ballooning the executable size.

David-JonesDVN commented:
In addition, I noticed several problems with WebView2:
There is a fourth assembly (the runtime) that needs to be included (two, actually. One for each CPU architecture) in the executable to run, which to be frank, borders on bloat, but that's Microsoft, I guess.
The library seems to keep a persistent cache in the working directory when the replay browser is used. Can it be disabled, maybe?
EDIT: You can also safely remove Microsoft.Web.WebView2.Wpf.dll for now. It should replace Microsoft.Web.WebView2.WinForms.dll in the next pull request.

DalekCraft2 commented:
Should I include all of the runtimes, or just the x64 one?

DalekCraft2 committed:
Use `.exe` files in root `thcrap` folder instead of the `thcrap/bin` folder

As per the recommendation of someone from the Touhou Patch Center Discord server.

David-JonesDVN commented:
Depends on whether we want to keep WebView2 at all.
If the cache folder can be disabled and we're willing to take the ~800KB size increase, I'd say include both (x86 and x64).
EDIT: I'm not sure about whether the 'Any CPU' option also allows for ARM architecture, but if it does, you may also include that.

DalekCraft2 commented:
I'm looking around online, and it seems that the cache can not really be disabled, but it can be cleared after the program closes, and it can also have its location changed. According to [this](https://stackoverflow.com/questions/66500241/webview2-site-not-loading/66501901#66501901), it can be set to the user's temp folder, which would get it out of the way of the working directory.

For embedding the runtimes, though, it seems it can not be done for some reason. At this point, with all of these files, it might be easier to ditch the single-file executable and simply distribute the program as a zip with the DLLs and the executable inside it (and maybe to add an auto-updater to make updating all of the files less of a pain).

Alternatively, we could switch back to .NET's `WebBrowser`, but that thing sucks.
As a second alternative, we could reconsider Costura.Fody and see whether it can include the runtimes. Using Costura.Fody would also make the project files less messy.

<sub>i recommend adding an update checker/auto-updater at some point regardless of what decision we make here, though</sub>

DalekCraft2 committed:
Find a much simpler way to change a file extension

DalekCraft2 committed:
Remove unnecessary EmbeddedResource tags

DalekCraft2 committed:
Remove WPF resource

DalekCraft2 committed:
Revert "Remove WPF resource"

This reverts commit 6e5099ea8a48551d4b341d187d0029dcda09bb9c.

DalekCraft2 committed:
Undo migration to WebView2 for now

DalekCraft2 commented:
I decided to just undo the migration to WebView2 for now because it's complicating things. Maybe the single-file publishing in .NET 6.0 will allow including the runtimes.

DalekCraft2 committed:
Consistent number style in Touhou game titles

DalekCraft2 commented:
You can test the build again, by the way.
